common/src/main/java/io/netty/util/concurrent/Futures.java

```diff
@@ -0,0 +1,208 @@
+     * @param mapper The function that will convert the result of this future into the result of the returned future.
+     * @param <R>    The result type of the mapper function, and of the returned future.
+     * @return A new future instance that will complete with the mapped result of this future.
+     */
+    public static <V, R> Future<R> map(Future<V> future, Function<V, R> mapper) {
+        Objects.requireNonNull(future, "The future cannot be null.");
+        Objects.requireNonNull(mapper, "The mapper function cannot be null.");
+        if (future.isSuccess()) {
+            return future.executor().submit(new CallableMapper<>(future, mapper));
+        }
+        Promise<R> promise = future.executor().newPromise();
+        future.addListener(new Mapper<>(promise, mapper));
+        promise.addListener(future, PromiseNotifier::propagateCancel);
+        return promise;
+    }
+
+    /**
+     * Creates a new {@link Future} that will complete with the result of the given {@link Future} flat-mapped through
+     * the given mapper function.
+     * <p>
+     * The "flat" in "flat-map" means the given mapper function produces a result that itself is a future-of-R, yet this
+     * method also returns a future-of-R, rather than a future-of-future-of-R. In other words, if the same mapper
+     * function was used with the {@link #map(Future, Function)} method, you would get back a {@code Future<Future<R>>}.
+     * These nested futures are "flattened" into a {@code Future<R>} by this method. Note that the future returned by
+     * this method is not the same instance as the one the mapper function returns. The reason is that this method needs
+     * to return immediately, but the mapper function cannot be applied before this future has completed.
+     * <p>
+     * Effectively, this method behaves similar to this serial code, except asynchronously and with proper exception
+     * and cancellation handling:
+     * <pre>{@code
+     * V x = future.sync().getNow();
+     * Future<R> y = mapper.apply(x);
+     * R result = y.sync().getNow();
+     * }</pre>
+     * <p>
+     * If the given future fails, then the returned future will fail as well, with the same exception.
+     * Cancellation of either future will cancel the other.
+     * If the mapper function throws, the returned future will fail, but the given future will be unaffected.
+     *
+     * @param mapper The function that will convert the result of this future into the result of the returned future.
+     * @param <R>    The result type of the mapper function, and of the returned future.
+     * @return A new future instance that will complete with the mapped result of this future.
+     */
+    public static <V, R> Future<R> flatMap(Future<V> future, Function<V, Future<R>> mapper) {
+        Objects.requireNonNull(future, "The future cannot be null.");
+        Objects.requireNonNull(mapper, "The mapper function cannot be null.");
```

Comment:
Are we changing our convention for NPE messages? It used to be:

```suggestion
        Objects.requireNonNull(future, "future");
        Objects.requireNonNull(mapper, "mapper");
```

Reply:
I like our old style more but I guess it's a matter of taste

Reply:
That's fine, but consistency is important, so if we're changing it then we have to change them all.